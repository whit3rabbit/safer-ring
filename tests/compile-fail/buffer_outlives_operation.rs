// This test should fail to compile because the buffer doesn't outlive the operation

use safer_ring::{Ring, PinnedBuffer};

fn main() {
    let ring = Ring::new(10).unwrap();
    
    {
        let buffer = PinnedBuffer::with_capacity(1024);
        // This should not compile - buffer lifetime is too short
        let _future = ring.read(0, buffer.as_mut_slice());
    } // buffer dropped while operation is pending
}