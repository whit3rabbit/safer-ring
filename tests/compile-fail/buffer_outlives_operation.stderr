error[E0597]: `ring` does not live long enough
  --> tests/compile-fail/buffer_outlives_operation.rs:11:23
   |
6  |     let ring = Ring::new(10).unwrap();
   |         ---- binding `ring` declared here
...
11 |         let _future = ring.read(0, buffer.as_mut_slice());
   |                       ^^^^ borrowed value does not live long enough
12 |     } // buffer dropped while operation is pending
13 | }
   | -
   | |
   | `ring` dropped here while still borrowed
   | borrow might be used here, when `ring` is dropped and runs the `Drop` code for type `Ring`

error[E0596]: cannot borrow `buffer` as mutable, as it is not declared as mutable
  --> tests/compile-fail/buffer_outlives_operation.rs:11:36
   |
11 |         let _future = ring.read(0, buffer.as_mut_slice());
   |                                    ^^^^^^ cannot borrow as mutable
   |
help: consider changing this to be mutable
   |
9  |         let mut buffer = PinnedBuffer::with_capacity(1024);
   |             +++
